File: Managed/UnrealSharp/UnrealSharp/Interop/FEditorDelegatesExporter.cs
using System.Runtime.InteropServices;

namespace UnrealSharp.Interop;

public struct FEditorDelegates
{
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate void FOnPIEEvent(NativeBool sessionEnded);
}

[NativeCallbacks]
public static unsafe partial class FEditorDelegatesExporter
{
    public static delegate* unmanaged<IntPtr, out FDelegateHandle, void> BindStartPIE;
    public static delegate* unmanaged<IntPtr, out FDelegateHandle, void> BindEndPIE;
    public static delegate* unmanaged<FDelegateHandle, void> UnbindStartPIE;
    public static delegate* unmanaged<FDelegateHandle, void> UnbindEndPIE;
}

File: Source/UnrealSharpCore/Private/Export/FEditorDelegatesExporter.cpp


#if WITH_EDITOR
#include "Editor.h"
#endif

#include "CSBindsManager.h"

using FPIEEvent = void(*)(bool);

DECLARE_UNREALSHARP_EXPORTER(FEditorDelegatesExporter)
{
	void BindEndPIE(FPIEEvent Delegate, FDelegateHandle* DelegateHandle)
	{
#if WITH_EDITOR
		*DelegateHandle = FEditorDelegates::EndPIE.AddLambda(Delegate);
#endif
	}

	void BindStartPIE(FPIEEvent Delegate, FDelegateHandle* DelegateHandle)
	{
#if WITH_EDITOR
		*DelegateHandle = FEditorDelegates::BeginPIE.AddLambda(Delegate);
#endif
	}

	void UnbindStartPIE(FDelegateHandle DelegateHandle)
	{
#if WITH_EDITOR
		FEditorDelegates::BeginPIE.Remove(DelegateHandle);
#endif
	}

	void UnbindEndPIE(FDelegateHandle DelegateHandle)
	{
#if WITH_EDITOR
		FEditorDelegates::EndPIE.Remove(DelegateHandle);
#endif
	}
	
	EXPORT_UNREALSHARP_FUNCTION(BindEndPIE)
	EXPORT_UNREALSHARP_FUNCTION(BindStartPIE)
	EXPORT_UNREALSHARP_FUNCTION(UnbindStartPIE)
	EXPORT_UNREALSHARP_FUNCTION(UnbindEndPIE)
}

